### ISSUE TITLE
are clipping options from leaflet implemented in folium?

### ISSUE THREAD

**Issue by jwhendy** [python-visualization/folium#752]
I'm completely new to `folium` and `leaflet` as of today, so maybe this is a silly question. I am plotting a `LineString` (though `Polygon` has the same behavior) on map and I noticed that I get visual artifacts when I scroll the map such that the perimeter is off the viewable area.

From googling around, I *think* it sounds and looks pretty much like [this issue](https://github.com/Leaflet/Leaflet/issues/2814).

Here's a simple version to illustrate:
```
import folium

lat, lon = (44, -93)
m = folium.Map(location=(lat, lon),
               zoom_start=12)
box = [[-93.2, 43.8], [-92.8, 43.8], [-92.8, 44.2], [-93.2, 44.2]]
line = {'type': 'Feature',
        'geometry': {'type': 'LineString',
                     'coordinates': coords }
        }

folium.GeoJson(line, name='line',
               style_function=lambda x: {'color': '#999999',
                                         'fill': '#999999',
                                         'opacity': 0.8,
                                         'fillOpacity': 0.5}).add_to(m)
m
```

This is what I get on `Chromium 62.0.3202.62` and `Firefox 56.0.1`:

![anim](https://user-images.githubusercontent.com/379357/31856730-05fc08e0-b68e-11e7-955b-877e8f1e5f45.gif)

Is this an instance of the clipping documented on the `leaflet` issue? If so, is there a way to get at `CLIP_PADDING` or `padding` as referenced over there?

**Comment by ocefpaf** [comment]
That clipping bug was fixed in leaflet a long time ago. Your issue is a malformed rectangle b/c you forgot to close it. This should work (note the repetition of the first point that closes the box):

```python
import folium

lat, lon = (44, -93)
m = folium.Map(
    location=(lat, lon),
    zoom_start=9
)

coords = [
    [-93.2, 43.8],
    [-92.8, 43.8],
    [-92.8, 44.2],
    [-93.2, 44.2],
    [-93.2, 43.8],
]

line = {
    'type': 'Feature',
    'geometry': {
        'type': 'LineString',
        'coordinates': coords
    }
}

folium.GeoJson(
    line,
    name='line',
    style_function=lambda x: {
        'color': '#999999',
        'fill': '#999999',
        'opacity': 0.8,
        'fillOpacity': 0.5}
).add_to(m)

m
```

To prevent errors like that you can use the `Rectangle` class (where you give the bounds and leaflet will do the rest):

```python
m = folium.Map(
    location=(lat, lon),
    zoom_start=9
)

bounds = [
    [44.2, -93.2],
    [43.8, -92.8],
]
    

folium.Rectangle(
    bounds,
    color='#999999',
    fill='#999999',
    opacity=0.8,
    fill_opacity=0.5,
).add_to(m)

m
```

**Comment by jwhendy** [comment]
Agreed that was old, and it looks like [this fixed](https://github.com/mapseed/platform/pull/405) the linked issue. I couldn't understand what that did, though, or if it's the new default or requires any action on my part.

I tried closing my box and saw no difference... turns out it was a variable name issue, which I just fixed. I was changing `box`, but my `GeoJson` was looking at `coords`, a spillover from it's previous def in my jupyter session...

I still get artifacts when I move your closed version off the screen; do you?
![2017-10-22_110508](https://user-images.githubusercontent.com/379357/31863846-2cf16806-b719-11e7-95d2-ce87ea24669a.png)

I wasn't able to use the `Rectangle`; do I need to import something differently? I'm sort of new to `python` in general, and I'm not always clear on how one imports. It [looks like](http://python-visualization.github.io/folium/docs-master/modules.html#folium.vector_layers.Rectangle) this is in `folium.vector_layers`, for example, but I can't import that. I get this using your code above: 
```
AttributeError: module 'folium' has no attribute 'Rectangle'
```

**Comment by jwhendy** [comment]
Figured it out. 0.5.0 release [is dated](https://pypi.python.org/pypi/folium) 2017-09-13; looks like `vector_layers` were [created or refactored](https://github.com/python-visualization/folium/commits/dd2425070572c31bdb7b3c2ae48719c547d77d4e/folium/vector_layers.py) on 2017-09-17. I installed from `git` and now I can use this.

My real shape is a complex polygon and I still show clipping/glitching with `GeoJson`, but `vector_layers.Polygon` is working great. Thanks!

**Comment by jwhendy** [comment]
Following back up on this. Was anyone able to replicate the `GeoJson` clipping, despite using a modern version (`0.5.0+44.g1053a02`) and closing the path with a 5th redundant point as shown above? I think this still looks like a bug (even though in real life I went with `vector_layers`).

**Comment by ocefpaf** [comment]
> I think this still looks like a bug (even though in real life I went with `vector_layers`).

I recommend you to put a minimum HTML example together and file a bug in leaflet's repo. Folium just wraps it, we don't do actually any HTML rendering directly.

**Comment by jwhendy** [comment]
@ocefpaf sounds great, and will do. I'll look into getting the `folium` object into raw html so I can provide it to them in that format.

**Comment by jwhendy** [comment]
Bah. Well, as is not *that* surprising (as it happens to me often), in the process of clarifying my thoughts for an issue at leaflet, I had a sudden burst of insight: *what if the issue is the fact that I defined a `LineString` and then said to give it a `fill`*?

Reviewing some documentation I'd run into when starting to play with `folium`, I found [this](https://macwright.org/2015/03/23/geojson-second-bite#linestrings) on that geometry:

>Points and lines are also similar in that they **have no area**: there is no inside or outside.

Huh. Well that's interesting. Only one way to find out how this plays out in `folium/leaflet` and that was to specify `'type': 'Polygon'` instead, which I also learned requires one more level of list nesting. With the original example above, I just changed it to this:

```
coords = [[
    [-93.2, 43.8],
    [-92.8, 43.8],
    [-92.8, 44.2],
    [-93.2, 44.2],
    [-93.2, 43.8],
]]
```

It worked! I don't get that triangle artifact with this method. Pretty sure it's a fluke that you *can* specify a `fill` for a line, but perhaps that's an abuse of the shape's intent and it just doesn't handle it well?

Mystery solved, programmer's angst resolved, thorn in brain removed... I can now sleep well tonight, and thanks once again for the assistance/patience! Hopefully this trail of breadcrumbs helps someone down the road.
